packages/gatsby/src/query/index.js

```diff
@@ -161,6 +161,20 @@ const groupQueryIds = queryIds => {
 const processQueries = async (queryJobs, activity) => {
   const queue = queryQueue.createBuildQueue()
   await queryQueue.processBatch(queue, queryJobs, activity)
+  if (process.env.GATSBY_PAGE_BUILD_ON_DATA_CHANGES) {
+    const { pages } = store.getState()
+    const { pageData } = readState()
+
+    if (pageData) {
+      readState().pageData.forEach((value, key) => {
```

Comment:
Maybe we can pass the cached pageData from the build function here.

Reply:
Can eliminate the explicit `if` branch by:

```suggestion
      pageData?.forEach((value, key) => {
```

Reply:
I'm getting a `no-unused-expressions` lint error 🤔

Reply:
I would argue to keep original code here (maybe just lose `readState().pageData` to `pageData` - the optional chaining here makes it more difficult to read what's going on ;)

Reply:
@StuartRayson That's because we're currently using an old version of ESLint that doesn't understand optional chaining